File: 프로그래머스/1/468370. 중요한 단어를 스포 방지/중요한 단어를 스포 방지.java
import java.util.HashSet;
import java.util.Set;

public class Solution {

    public int solution(String message, int[][] spoiler_ranges) {
        Set<String> spoilerWords = new HashSet<>();
        Set<String> normalWords = new HashSet<>();

        boolean[] isSpoiler = new boolean[message.length()];
        for (int i = 0; i < spoiler_ranges.length; i++) {
            for (int j = spoiler_ranges[i][0]; j <= spoiler_ranges[i][1]; j++) {
                isSpoiler[j] = true;
            }
        }

        StringBuilder sb = new StringBuilder();
        boolean flag = false;

        for (int i = 0; i < message.length(); i++) {
            char c = message.charAt(i);

            if (c == ' ') {
                if (sb.length() > 0) {
                    String word = sb.toString();
                    if (flag && !normalWords.contains(word)) {
                        spoilerWords.add(word);
                    } else {
                        normalWords.add(word);
                    }
                    sb.setLength(0);
                    flag = false;
                }
            } else {
                sb.append(c);
                if (isSpoiler[i]) {
                    flag = true;
                }
            }
        }

        if (sb.length() > 0) {
            String word = sb.toString();
            if (flag && !normalWords.contains(word)) {
                spoilerWords.add(word);
            } else {
                normalWords.add(word);
            }
        }

        int answer = 0;
        for (String word : spoilerWords) {
            if (!normalWords.contains(word)) {
                answer++;
            }
        }

        return answer;
    }
}

File: 프로그래머스/1/468370. 중요한 단어를 스포 방지/중요한 단어를 스포 방지.py
def solution(message, spoiler_ranges):
    spoiler_words = set()
    normal_words = set()

    is_spoiler = [False] * len(message)
    for start, end in spoiler_ranges:
        for i in range(start, end + 1):
            is_spoiler[i] = True

    sb = []
    flag = False

    for i, c in enumerate(message):
        if c == ' ':
            if sb:
                word = ''.join(sb)
                if flag and word not in normal_words:
                    spoiler_words.add(word)
                else:
                    normal_words.add(word)
                sb = []
                flag = False
        else:
            sb.append(c)
            if is_spoiler[i]:
                flag = True

    if sb:
        word = ''.join(sb)
        if flag and word not in normal_words:
            spoiler_words.add(word)
        else:
            normal_words.add(word)

    answer = 0
    for word in spoiler_words:
        if word not in normal_words:
            answer += 1

    return answer